### PR TITLE
Add rule Scope configuration option to rules-based sampler

### PR DESCRIPTION
**Pull request by isnotajoke**
## Which problem is this PR solving?

We'd like to be able to apply all of the conditions in a rules-based sampler rule to a single span. In other words, we'd like a rule to be considered matched only if *all* of the conditions in that rule are satisfied by a single span in the trace. This allows us to more accurately make per-service sampling decisions in a dataset written to by multiple services.

- Closes see #420

## Short description of the changes

At a high level, we add a new `Scope` configuration option to the rule definition. By default (and if unspecified) it is "trace". If "span", all conditions in the rule must be satisfied together by some span in a trace. This is a subtle but important difference with the default behavior, effectively making a rule more strict than it would otherwise be.

`Scope` introduces some complexity to the rule based sampler. While many of the operations performed within the 
`GetSampleRate` method (extracting a value, checking for a match) are similar between the two approaches, the shape of the logic to compute a match is totally different when `Scope` is 'trace' than when it is 'span'. To avoid duplication, we refactor the common methods into private functions; this also makes the logic to loop through & compute a match a little easier to understand at a glance.

I added a basic spec to cover the new behavior.

### CODE REVIEW

File: config/sampler_config.go [hunk 1]
```diff
@@ -62,6 +62,7 @@ type RulesBasedSamplerRule struct {
 	SampleRate int
 	Sampler    *RulesBasedDownstreamSampler
 	Drop       bool
+	MatchSpan  bool
```
Comment (isnotajoke) [inline config/sampler_config.go:65]: Probably need to add this to the documentation + example rule configurations.

**Review by MikeGoldsmith — CHANGES_REQUESTED** [review]
Hey @isnotajoke - thank you for putting this together. Sorry for the delay in getting it reviewed.

Overall, I really like what you've done and think it's a great solution. I've left a suggestion for changing the configuration option - let me know you thoughts.

**Comment by isnotajoke** [comment]
@MikeGoldsmith makes sense to me. I'll go ahead and make that change.

**Comment by isnotajoke** [comment]
We're planning to test this PR in our infra this week. Will update with any concerns/follow up tweaks after that, and then it should be good for final review.

File: sample/rules.go [hunk 2]
```diff
@@ -67,10 +67,17 @@ func (s *RulesBasedSampler) GetSampleRate(trace *types.Trace) (rate uint, keep b
...
 			matched = ruleMatchesSpanInTrace(trace, rule, s.Config.CheckNestedFields)
-		} else {
+		case "trace", "":
 			matched = ruleMatchesTrace(trace, rule, s.Config.CheckNestedFields)
+		default:
+			logger.WithFields(map[string]interface{}{
+				"rule_name": rule.Name,
+				"scope":     rule.Scope,
+			}).Logf("invalid scope %s given for rule: %s", rule.Scope, rule.Name)
+			matched = false
```
Comment (MikeGoldsmith) [inline sample/rules.go:80]: I think we should default to send traces for misconfigured rules.

```suggestion
			matched = true
```

**Review by MikeGoldsmith — DISMISSED** [review]
Looks great, thanks @isnotajoke.

Yes, please do test and report back any findings. We'll hold back on merging this for your feedback.

File: rules_complete.toml [hunk 3]
```diff
@@ -242,6 +242,26 @@ SampleRate = 1
...
+	[[dataset4.rule]]
+		name = "sample traces originating from a service"
+		# if scope is set to "span", a single span in the trace must match
+		# *all* of the conditions associated with this rule for the rule to
+		# apply to the trace.
+		#
+		# this is especially helpful when sampling a dataset written to
+		# by multiple services that call one another in normal operation –
+		# you can set MatchSpan to true to attribute traces to an origin
+		# service in a way that would be difficult without it.
```
Comment (ajvondrak) [inline rules_complete.toml:254]: Missed this comment when changing from `MatchSpan` to `Scope`. 👀

File: rules_complete.toml [hunk 4]
```diff
@@ -242,6 +242,26 @@ SampleRate = 1
...
 
+	[[dataset4.rule]]
+		name = "sample traces originating from a service"
+		# if scope is set to "span", a single span in the trace must match
+		# *all* of the conditions associated with this rule for the rule to
+		# apply to the trace.
+		#
+		# this is especially helpful when sampling a dataset written to
+		# by multiple services that call one another in normal operation –
+		# you can set MatchSpan to true to attribute traces to an origin
```
Comment (MikeGoldsmith) [inline rules_complete.toml:253]:
```suggestion
		# you can set Scope to 'span' to attribute traces to an origin
```

**Comment by MikeGoldsmith** [comment]
FYI the test failure is coming from [config_test](https://github.com/honeycombio/refinery/blob/main/config/config_test.go#L249) - you just need to update the number of rules in the dataset to be 5.

**Comment by MikeGoldsmith** [comment]
> We're planning to test this PR in our infra this week. Will update with any concerns/follow up tweaks after that, and then it should be good for final review.

Hey @isnotajoke - just wanted to check in to see how your testing went?

**Comment by isnotajoke** [comment]
@MikeGoldsmith still ongoing. Should be complete this week.

File: sample/rules.go (hunk 2 above)
Comment (isnotajoke) [reply sample/rules.go:80]: Makes sense. Will update.

**Comment by MikeGoldsmith** [comment]
Hey @isnotajoke - just wanted to check in to see if you've completed your internal testing?

We're eager to move forward with this feature and value your feedback.

**Comment by MikeGoldsmith** [comment]
Hey @isnotajoke - I'm going to look to move ahead with this change this week. Let me know if you have any additional feedback.

**Comment by isnotajoke** [comment]
@MikeGoldsmith sorry for the radio silence here. We completed our testing and didn't see any issues. No concerns with moving ahead.

**Pull request merged**